paho/client.go

```diff
@@ -0,0 +1,378 @@
+}
+
+// NewClient is used to create a new instance of an MQTT client. It
+// returns a pointer to the new client instance and an error.
+// opts is a variadic of functions that take a pointer to a Client and
+// returns an error. These functions are used to modify the properties
+// of the client (such as setting the PingHandler, Router, etc), a
+// selection of such functions are provided in this library.
+func NewClient(opts ...func(*Client) error) (*Client, error) {
+	debug.Println("Creating new client")
+	c := &Client{
+		Stop: make(chan struct{}),
+	}
+
+	debug.Println("Setting options")
+	for _, opt := range opts {
+		if err := opt(c); err != nil {
+			return nil, err
+		}
+	}
+
+	if c.PingHandler == nil {
+		c.PingHandler = &PingHandler{
+			pingFailHandler: func(e error) {
+				c.Error(e)
+			},
+		}
+	}
+
+	if c.Persistence == nil {
+		c.Persistence = &noopPersistence{}
+	}
+
+	if c.MIDs == nil {
+		c.MIDs = &MIDs{index: make(map[uint16]*CPContext)}
+	}
+
+	if c.PacketTimeout == 0 {
+		c.PacketTimeout = 10 * time.Second
+	}
+
+	return c, nil
+}
+
+// Connect is used to connect the client to a server. It presumes that
+// the Client instance already has a working network connection.
+// The function takes a pre-prepared Connect packet, and uses that to
+// establish an MQTT connection. Assuming the connection completes
+// successfully the rest of the client is initiated and the Connack
+// returned. Otherwise the failure Connack (if there is one) is returned
+// along with an error indicating the reason for the failure to connect.
+func (c *Client) Connect(cp *p.Connect) (*p.Connack, error) {
+	debug.Println("Connecting")
+	c.Lock()
+	defer c.Unlock()
+
+	if c.Conn == nil {
+		return nil, fmt.Errorf("Client has no connection")
+	}
+
+	cp.ProtocolName = "MQTT"
+	cp.ProtocolVersion = 5
+
+	debug.Println("Sending CONNECT")
+	if err := cp.Send(c.Conn); err != nil {
+		return nil, err
+	}
+
+	debug.Println("Waiting for CONNACK")
+	ca, err := p.ReadPacket(c.Conn)
+	if err != nil {
+		return nil, err
+	}
+
+	if ca.Type != p.CONNACK {
+		return nil, fmt.Errorf("Received %d instead of Connack", ca.Type)
+	}
+
+	if ca.Content.(*p.Connack).ReasonCode >= 0x80 {
+		debug.Println("Received an error code in Connack:", ca.Content.(*p.Connack).ReasonCode)
+		return ca.Content.(*p.Connack), fmt.Errorf("Failed to connect to server: %s", ca.Content.(*p.Connack).Reason())
+	}
+
+	debug.Println("Received CONNACK, starting PingHandler and Incoming")
+	c.Workers.Add(2)
+	go c.PingHandler.Start(c.Conn, time.Duration(cp.KeepAlive)*time.Second)
+	go c.Incoming()
+
+	return ca.Content.(*p.Connack), nil
+}
+
+// Incoming is the Client function that reads and handles incoming
+// packets from the server. The function is started as a goroutine
+// from Connect(), it exits when it receives a server initiated
+// Disconnect, the Stop channel is closed or there is an error reading
+// a packet from the network connection
+func (c *Client) Incoming() {
```

Comment:
Why put it as a public function?
It seems it is just an internal implementation detail, and calling it twice might create havoc.